ddtrace/tracer/textmap.go

```diff
@@ -487,3 +501,292 @@ func (*propagatorB3) extractTextMap(reader TextMapReader) (ddtrace.SpanContext,
+// hex-encoded digits into a 64-bit number.
+func parseTraceparent(ctx *spanContext, header string) error {
+	nonWordRegex := "\\W_"
+	header = strings.ToLower(strings.Trim(header, "\t -"))
+	if len(header) == 0 {
+		return ErrSpanContextNotFound
+	}
+	if len(header) != 55 {
+		return ErrSpanContextCorrupted
+	}
+	parts := strings.Split(header, "-")
+	if len(parts) != 4 {
+		return ErrSpanContextCorrupted
+	}
+	version := strings.Trim(parts[0], nonWordRegex)
+	if len(version) != 2 {
+		return ErrSpanContextCorrupted
+	}
+	if v, err := strconv.ParseUint(version, 16, 64); err != nil || v == 255 {
+		return ErrSpanContextCorrupted
+	}
+	// parsing traceID
+	fullTraceID := strings.Trim(parts[1], nonWordRegex)
+	if len(fullTraceID) != 32 {
+		return ErrSpanContextCorrupted
+	}
+	// checking that the entire TraceID is a valid hex string
+	if ok, err := regexp.MatchString("^[a-f0-9]+$", fullTraceID); !ok || err != nil {
+		return ErrSpanContextCorrupted
+	}
+	var err error
+	if ctx.traceID, err = strconv.ParseUint(fullTraceID[16:], 16, 64); err != nil {
+		return ErrSpanContextCorrupted
+	}
+	if ctx.traceID == 0 {
+		if strings.Trim(fullTraceID[:16], "0") == "" {
+			return ErrSpanContextNotFound
+		}
+	}
+	// parsing spanID
+	spanID := strings.Trim(parts[2], nonWordRegex)
+	if len(spanID) != 16 {
+		return ErrSpanContextCorrupted
+	}
+	if ok, err := regexp.MatchString("[a-f0-9]+", spanID); !ok || err != nil {
+		return ErrSpanContextCorrupted
+	}
+	if ctx.spanID, err = strconv.ParseUint(spanID, 16, 64); err != nil || ctx.spanID == 0 {
+		return ErrSpanContextNotFound
+	}
+	// parsing flags
+	flags := parts[3]
+	f, err := strconv.ParseInt(flags, 16, 8)
+	if err != nil {
+		return ErrSpanContextCorrupted
+	}
+	// setting trace-id to be used for span context propagation
+	setPropagatingTag(ctx, w3cTraceIDTag, fullTraceID)
+	ctx.setSamplingPriority(int(f)&0x1, samplernames.Unknown)
+	return nil
+}
+
+// parseTracestate attempts to parse tracestateHeader which is a list
+// with zero or more comma-separated (,) list-members.
+// An example value would be: `vendorname1=opaqueValue1,vendorname2=opaqueValue2,dd=s:1;o:synthetics`,
+// Where `dd` list contains values that would be in x-datadog-tags as well as those needed for propagation information.
+// The keys to the “dd“ values have been shortened as follows to save space:
+// `sampling_priority` = `s`
+// `origin` = `o`
+// `_dd.p.` prefix = `t.`
+func parseTracestate(ctx *spanContext, header string) error {
+	// if multiple headers are present, they must be combined and stored
+	setPropagatingTag(ctx, tracestateHeader, header)
+	list := strings.Split(strings.Trim(header, "\t "), ",")
+	for _, s := range list {
+		if !strings.HasPrefix(s, "dd=") {
+			continue
+		}
+		dd := strings.Split(s[len("dd="):], ";")
+		tags := make(map[string]string)
+		for _, val := range dd {
+			x := strings.SplitN(val, ":", 2)
+			if len(x) != 2 {
+				continue
+			}
+			tags[x[0]] = x[1]
+		}
+		for k, v := range tags {
```

Comment:
Do we need to iterate twice and build the `tags` map?

Can we eliminate `tags` and the second loop here by just saying this?
```suggestion
			k, v := x[0], x[1]
```

Reply:
This is the approach @katiehockman suggested. Take a look here 
https://github.com/DataDog/dd-trace-go/pull/1616#discussion_r1049991385.
Could rewrite it too 

```
dd := strings.Split(s[len("dd="):], ";")
for _, val := range dd {
	x := strings.SplitN(val, ":", 2)
	if len(x) != 2 {
		continue
	}
	k, v := x[0], x[1]
	if k == "o" {
		ctx.origin = v
	} else if k == "s" {
		p, err := strconv.Atoi(v)
		if err != nil {
			// if the tracestate priority is absent, relying on traceparent value
			continue
		}
		flagPriority, _ := ctx.samplingPriority()
		if (flagPriority == 1 && p > 0) || (flagPriority == 0 && p <= 0) {
			ctx.setSamplingPriority(p, samplernames.Unknown)
		}
	} else if strings.HasPrefix(k, "t.") {
		k = k[len("t."):]
		v = strings.ReplaceAll(v, "~", "=")
		setPropagatingTag(ctx, "_dd.p."+k, v)
	}
}
```

Reply:
The above looks good.

Unless there is a reason to construct the map, we can avoid the extra allocations and extra loop.